content/posts/numpy/numpy-rng/index.md

```diff
@@ -0,0 +1,144 @@
+
+In short:
+
+- Instead of using `np.random.seed`, which reseeds the already created global NumPy RNG, and then using `np.random.*` functions, you should create a new RNG.
+- You should create an RNG at the beginning of your script (with your own seed if you want reproducibility) and use this RNG in the rest of the script.
+
+To create a new RNG you can use the `default_rng` function as illustrated in the [introduction of the random module documentation](https://numpy.org/doc/stable/reference/random/index.html):
+
+```python
+import numpy as np
+
+rng = np.random.default_rng()
+rng.random()  # generate a floating point number between 0 and 1
+```
+
+If you want to use a seed for reproducibility, [the NumPy documentation recommends](https://numpy.org/doc/stable/reference/random/index.html#quick-start) using a very large and unique number to have a different seed than anyone else. Another reason is that using a small number for the seed could lead to a bad first internal state of the RNG leading to a sequence of numbers that would not look like a random sequence, although `default_rng` now takes care of creating a good internal state from a small number. To obtain a good and unique seed you can rely on the [secrets module](https://docs.python.org/3/library/secrets.html). The following code will return a 128 bit number that you can then use as a seed:
+
+```python
+import secrets
+
+secrets.randbits(128)
+```
+
+When running this code I get `65647437836358831880808032086803839626` for the number to use as my seed. This number is randomly generated so you need to copy paste the value that is returned by `secrets.randbits(128)` otherwise you will have a different seed each time you run your code and thus break reproducibility:
+
+```python
+import numpy as np
+
+seed = 65647437836358831880808032086803839626
+rng = np.random.default_rng(seed)
+rng.random()
+```
+
+The reason for seeding your RNG only once (and passing that RNG around) is that with a good RNG such as the one returned by `default_rng` you will be ensured good randomness and independence of the generated numbers. However, if not done properly, using several RNGs (each one created with its own seed) might lead to streams of random numbers that are less independent than the ones created from the same seed. That being said, [as explained by Robert Kern](https://github.com/numpy/numpy/issues/15322#issuecomment-573890207), with the RNGs and seeding strategies introduced in NumPy 1.17, it is considered fairly safe to create RNGs using system entropy, i.e. using `default_rng(None)` multiple times. However as explained later be careful when running jobs in parallel and relying on `default_rng(None)`. Another reason for seeding your RNG only once is that obtaining a good seed can be time consuming. Once you have a good seed to instantiate your generator, you might as well use it.
```

Comment:
Independence of generated numbers is a property of the algorithm, for any given seed?

Reply:
Yes for a given seed, the algorithm produces independent random numbers. For two different seeds you do not know whether the number generated by the different seeds conserve this property. For instance your second seed could lead to the internal state that you obtain after 3 calls to the RNG instantiated with the first seeds. Then you will have an overlap of the values generated by the two RNGs. Do you want me to clarify this?

Reply:
Personally, I find these subtleties interesting. Perhaps footnotes? But I leave it up to you.